integration: add WatchFromCurrentRevision test

gyuho opened the pull request.

gyuho commented:
/cc @xiang90 @heyitsanthony 

Thanks

heyitsanthony commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +196,157 @@ func TestV3DeleteRange(t *testing.T) {
+		value         string
+		watchRequests []pb.WatchRequest
+
+		eventNum int
```
is this ever different from len(events)?

gyuho replied:
Right. Will remove this.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +196,157 @@ func TestV3DeleteRange(t *testing.T) {
 }
+
+// TestV3Watch tests Watch APIs.
+func TestV3Watch(t *testing.T) {
```
TestV3WatchFromCurrentRevision

xiang90 replied:
Let's keep this test simple. Move the unsynced one into another test.

gyuho replied:
Ok will do. Thanks,

On Thu, Jan 14, 2016 at 10:15 AM, Xiang Li <notifications@github.com> wrote:

> In integration/v3_grpc_test.go
> <https://github.com/coreos/etcd/pull/4214#discussion_r49762181>:
>
> > @@ -193,3 +196,157 @@ func TestV3DeleteRange(t *testing.T) {
> >  		clus.Terminate(t)
> >  	}
> >  }
> > +
> > +// TestV3Watch tests Watch APIs.
> > +func TestV3Watch(t *testing.T) {
>
> Let's keep this test simple. Move the unsynced one into another test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4214/files#r49762181>.
>



-- 
Sincerely,
Gyu-Ho Lee

heyitsanthony commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +196,157 @@ func TestV3DeleteRange(t *testing.T) {
+
+		if len(tt.initialKeys) > 0 {
+			// give time to run syncWatchers
+			time.Sleep(100 * time.Millisecond)
```
why does this sleep? shouldn't wStream.Recv() block until the watch events get sent out?

gyuho replied:
You are right. Will remove this when separating `unsynced` part to another test.

heyitsanthony commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +196,157 @@ func TestV3DeleteRange(t *testing.T) {
+					}
+				}
+
+				if !reflect.DeepEqual(copied, tt.events) {
```
if all the events don't appear in a single response, then shouldn't this check go outside the `for` loop?

gyuho replied:
Correct. This was wrong trying to test both synced and unsynced. Will address in the next commit.
Thanks,

gyuho commented:
@heyitsanthony @xiang90 Just split it into two tests. PTAL. Thanks!

gyuho commented:
Locally tested and worked.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+		watchRequests []pb.WatchRequest
+		events        []storagepb.Event
+	}{
+		// watch the key
```
shall we also test the case when watching a non-matching key like prefix one?

xiang90 replied:
I feel we should also test the case that multiple events can be delivered to watcher in order.

gyuho replied:
Ok will add them. Thanks,

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+// TestV3WatchFromCurrentRevision tests Watch APIs from current revision.
+func TestV3WatchFromCurrentRevision(t *testing.T) {
+	tests := []struct {
+		initialKeys   []string
```
move initialValue of of the tests struct since they are the same

xiang90 replied:
I feel we should call this putKeys rather than initialKeys. We first do the watch, not initialize key space.

gyuho replied:
Sure, will change.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+		initialKeys   []string
+		initialValue  string
+		watchRequests []pb.WatchRequest
+		events        []storagepb.Event
```
events -> wevents (want events). also put a blank line between the input and expected output.

xiang90 replied:
we should check the entire response rather than just the received events (we should check headers)

gyuho replied:
Ok this is a good place to test ResponseHeader. Will add.

xiang90 replied:
right. we want to test if the whole API works as expected. Header is part of that.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+	tests := []struct {
+		initialKeys   []string
+		initialValue  string
+		watchRequests []pb.WatchRequest
```
why do we need to make this a slice? there is always only one watch request.

xiang90 replied:
we probably should test multiple concurrent watchers in another test.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+		wAPI := pb.NewWatchClient(clus.RandConn())
+		wStream, err := wAPI.Watch(context.TODO())
+		if err != nil {
+			t.Errorf("#%d: wAPI.Watch error: %v", i, err)
```
you should fatal here. the test cannot continue if you cannot get the stream

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+
+		for _, r := range tt.watchRequests {
+			if err := wStream.Send(&r); err != nil {
+				t.Errorf("#%d: wStream.Send error: %v", i, err)
```
same here. fatal instead.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+		for _, k := range tt.initialKeys {
+			reqput := &pb.PutRequest{Key: []byte(k), Value: []byte(tt.initialValue)}
+			if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+				t.Errorf("#%d: couldn't put key (%v)", i, err)
```
same here. fatal please.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+		cnt := 0
+		for cnt != len(tt.events) {
+			resp, err := wStream.Recv()
+			if err == io.EOF {
```
we never close the stream. io.EOF should be unexpected rather than just break out the loop.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+				break
+			}
+			if err != nil {
+				t.Errorf("#%d: wStream.Recv error: %v", i, err)
```
fatal here.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+
+			switch {
+			case resp.Created:
+				t.Logf("#%d: watcher created: id %08x\n", i, resp.WatchId)
```
remove this unnecessary log during testing.

xiang90 replied:
we should check we receive exact one create each time and the received events have the same watch ID as the one in create response.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+			}
+		}
+
+		copied := []storagepb.Event{}
```
not sure why do we want to copy stuff around. can you explain?

gyuho replied:
`Recv` returns `[]*Event` where I define test cases with `[]Event`. So I made one same to another.

xiang90 replied:
Can we define test case with []*Event. So we do not need to copy logic.

gyuho replied:
Sure.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+			t.Errorf("#%d: got = %+v, want = %+v", i, copied, tt.events)
+		}
+
+		for _, ev := range copied {
```
remove unnecessary logging

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
+}
+
+// TestV3WatchUnsynced tests Watch APIs when watchers are unsynced.
+func TestV3WatchUnsynced(t *testing.T) {
```
test failed on this one. Let's move this one to another pr and focus on merging the sync one first.

gyuho replied:
Ok will separate. Thanks,

gyuho commented:
@xiang90  Just updated addressing your feedback.

1. removed unsynced tests
2. removed unnecessary logs
3. fatal when watch fails
4. make the wevents type equivalent (no need to copy)


TODO:

1. fix 0 revision failures
2. better ResponseHeader test
3. test multiple concurrent watchers in another test
4. added a test for synced, multiple events
5. test the same patterns for unsynced

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,135 @@ func TestV3DeleteRange(t *testing.T) {
+		watchRequest   *pb.WatchRequest
+		headerRevision int64
+		createdN       int
+		wevents        []*storagepb.Event
```
let's change wevents to wresp then we can do deep equal with received response.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,135 @@ func TestV3DeleteRange(t *testing.T) {
+		cnt := 0
+		for cnt != len(tt.wevents) {
+			resp, err := wStream.Recv()
+			if err == io.EOF {
```
why not just remove line 281? we will catch this on line 284.

gyuho replied:
Sure, will remove.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,135 @@ func TestV3DeleteRange(t *testing.T) {
+
+			switch {
+			case resp.Created:
+				createdWatchId = resp.WatchId
```
well. if the assumption is that only one watcher should be created, then we should remove these `cnt` stuff. If the assumption is multiple watchers might be created, then line 298 can overwrite the previous ID.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,135 @@ func TestV3DeleteRange(t *testing.T) {
+			[]string{testKey},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Key: []byte(testKey2)}},
+			1,
+			0,
```
why this is 0? we should expect a watcher creation as long as you send a watch create request, no?

gyuho replied:
Yes that's correct. Will fix. somehow it wasn't returning CreatedRequest. Need to investigate what I am doing wrong.

Thanks,

gyuho replied:
Oh I set up the `for loop` condition wrong. Will fix soon.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,132 @@ func TestV3DeleteRange(t *testing.T) {
+
+// TestV3WatchFromCurrentRevision tests Watch APIs from current revision.
+func TestV3WatchFromCurrentRevision(t *testing.T) {
+	testKey, testKeyL, testKey2 := "foo", "fooLong", "helloworld"
```
this is pretty confusing. I feel we should just embed the key into the test cases. Then it is much more readable.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,132 @@ func TestV3DeleteRange(t *testing.T) {
+
+			if resp.Header == nil {
+				t.Fatalf("#%d: unexpected nil resp.Header", i)
+			} else {
```
remove this else. we fatal at line 285.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,132 @@ func TestV3DeleteRange(t *testing.T) {
+			case resp.Created:
+				createdWatchId = resp.WatchId
+				createdCnt++
+				continue
```
remove this continue.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,132 @@ func TestV3DeleteRange(t *testing.T) {
+		watchRequest   *pb.WatchRequest
+		headerRevision int64
+		createdN       int
+		wresp          []*storagepb.Event
```
i mean change []*storagepb.Event -> *WatchRespons

gyuho replied:
Ok will change.

gyuho commented:
@xiang90  Just updated. PTAL

TODO:

1. fix 0 revision failures
2. better ResponseHeader test. Fill in in `WatchResponse`.
3. test multiple concurrent watchers in another test
4. added a test for synced, multiple events (need Txn)
5. test the same patterns for unsynced

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+	tests := []struct {
+		putKeys        []string
+		watchRequest   *pb.WatchRequest
+		headerRevision int64
```
can we remove this headerRevision? this is already included in wresp, no? Also please put a blank line above wresp to separate the input and expect values.

gyuho replied:
Sure. Will change. Thanks!

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+		createdWatchId, isCreated := 0, false
+
+		cnt := 0
+		for cnt != len(tt.wresp.Events)+1 {
```
This is incorrect. we should expect exact 1 + len(puts) watch response. one for the creation, the others for the put requests. It has nothing to do with number of events in response as far as i can see.

gyuho replied:
You are right. Will fix it with `1 + len(putKeys)`. Thanks,

xiang90 replied:
this is only for the matching case. for non matching case, we should expect just the creation. Also the current test is also incorrect. we did not ensure that there is no more responses coming through the watcher. 

A better way is to put all expected response into wresp and then loop over that slice to match what we receive from the watcher. Then we need to try to receive again to ensure there is no more (do this in another routine and close the stream after a timeout).

xiang90 replied:
let's range over the wresps slice and compare them with the received one.

gyuho replied:
Ok that sounds better. Let me address that.

gyuho replied:
Yeah I was wrong. We should make it slice. Working on it. Thanks,

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+		evs := []*storagepb.Event{}
+		createdWatchId, isCreated := 0, false
+
+		cnt := 0
```
add w prefix to indicate this is the expected value. cnt -> wcnt

xiang90 commented:
@gyuho 

> better ResponseHeader test. Fill in in WatchResponse.

Why we cannot do it in this commit? I think we should do it since we want to check the entire response in this test.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+			switch {
+			case resp.Created:
+				createdWatchId = int(resp.WatchId)
+				if isCreated {
```
move this before line 293. or the watchId might get overwritten.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+				t.Logf("[TODO - skip for now] #%d: resp.Header.Revision got = %d, want = %d", i, resp.Header.Revision, tt.headerRevision)
+			}
+
+			switch {
```
we do not really need this. when we receive the nth response we should compare it with nth wresp (change wresp to a slice).

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+		putKeys        []string
+		watchRequest   *pb.WatchRequest
+		headerRevision int64
+		wresp          *pb.WatchResponse
```
change this to wreps []*pb.WatchResponse

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
+			[]string{"foo"},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Key: []byte("foo")}},
+			1,
+			&pb.WatchResponse{
```
change this to a slice. add the first expected create response to here.

gyuho replied:
Ok I will change it to slice.

gyuho commented:
@xiang90 We can match Revision in the header, but raftterm, memberid, clusterid seem hard to predict. Do you have any suggestion to test ResponseHeader?

xiang90 commented:
@gyuho We can ignore them for now. Do not use deep-equal then, we can compare the necessary ones manually.

gyuho commented:
@xiang90 Ok sounds good. Thanks,

gyuho commented:
@xiang90 Just changed it to `wresps` with other small changes based on your feedback.
It still checks only `Revision` in `ResponseHeader`.

Now TODO:

1. fix 0 revision failures
2. better `WatchResponse`'s `Header` comparison
3. test multiple concurrent watchers 
4. test synced, multiple events (need Txn)
5. test unsynced

Thanks!

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+				t.Errorf("#%d-%d: resp.Events got = %+v, want = %+v", i, j, evs, wresp.Events)
+			}
+		}
+
```
create another go routine here to try receive again. then we can make sure there is no more response are coming from the watch stream.

gyuho replied:
Ok, forgot to check that part. Will add. Thanks,

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+				t.Errorf("#%d-%d: wStream.Recv error: %v", i, j, err)
+			}
+
+			if j == 0 {
```
if resp.Create {
}

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+		for j, wresp := range tt.wresps {
+			resp, err := wStream.Recv()
+			if err != nil {
+				t.Errorf("#%d-%d: wStream.Recv error: %v", i, j, err)
```
#%d.%d instead of #%d-%d

gyuho replied:
Ok

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+					evs = append(evs, r)
+				}
+			}
+			if !reflect.DeepEqual(evs, wresp.Events) {
```
can we just do reflect.DeepEqual(resp.Events, wresp.Events)?

gyuho replied:
Just response from grpc seems to include `r.Kv` nil events, so the `reflect.Equal` fails.

xiang90 replied:
not sure i understand this. but we can change the events in wresp to match it, no?

gyuho replied:
Sure, let me try that. It will be cleaner.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+			if wresp.Created != resp.Created {
+				t.Errorf("#%d-%d: resp.Created got = %v, want = %v", i, j, resp.Created, wresp.Created)
+			}
+			if j != 0 && resp.Created {
```
we do not need this. if we only have one create in wresps, more creation will be detected automatically.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+				t.Errorf("#%d-%d: Watch created more than once", i, j)
+			}
+
+			if len(resp.Events) != len(wresp.Events) {
```
we do not need this compare. deepequal will compare the length.

gyuho commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,165 @@ func TestV3DeleteRange(t *testing.T) {
+				t.Errorf("#%d.%d: resp.Events got = %+v, want = %+v", i, j, evs, wresp.Events)
+			}
+		}
+
```
these goroutines leak, because they are running inside `for` loop. Test will fail from the goroutine leak check.

@xiang90 Any suggestion?

xiang90 replied:
you need to close the watch stream to stop the routine.

gyuho replied:
Oh got it. Thanks! Will fix!

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +194,156 @@ func TestV3DeleteRange(t *testing.T) {
+			}
+
+			if resp.Header == nil {
+				t.Errorf("#%d-%d: unexpected nil resp.Header", i, j)
```
we need to fatal here. if header is nil, the test should not continue or it panics when access header.revision.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,165 @@ func TestV3DeleteRange(t *testing.T) {
+		case nr := <-rCh:
+			t.Errorf("#%d: unexpected response is received %+v", i, nr)
+		case <-time.After(3 * time.Second):
+		}
```
here, close the wStream and try to receive nil resp from rCh

gyuho replied:
ok

gyuho commented:
@xiang90 I was assigning length 0 slice which is different than `nil`. 
Just fixed and now should work.

Thanks for feeedback! PTAL.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,159 @@ func TestV3DeleteRange(t *testing.T) {
+			},
+		},
+
+		// TODO: watch and receive multiple-events from synced (need Txn)
```
also we can test multiple puts to one watcher ([]string{"foo", "foo"}). multiple puts to one prefix watcher, etc..

gyuho replied:
Sure, will add.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,159 @@ func TestV3DeleteRange(t *testing.T) {
+			if wresp.Created != resp.Created {
+				t.Errorf("#%d.%d: resp.Created got = %v, want = %v", i, j, resp.Created, wresp.Created)
+			}
+			if j != 0 && resp.Created {
```
remove this. we do not need it. check my previous comment.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,159 @@ func TestV3DeleteRange(t *testing.T) {
+			}
+
+			if resp.Header == nil {
+				t.Errorf("#%d.%d: unexpected nil resp.Header", i, j)
```
fatalf. check my previous comment for this

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,159 @@ func TestV3DeleteRange(t *testing.T) {
+				t.Errorf("#%d.%d: resp.WatchId got = %d, want = %d", i, j, resp.WatchId, createdWatchId)
+			}
+
+			if len(resp.Events) != len(wresp.Events) {
```
we do not need this too. deepEqual is enough. check my previous comment.

xiang90 commented on integration/v3_grpc_test.go:
```diff
@@ -193,3 +195,159 @@ func TestV3DeleteRange(t *testing.T) {
+			t.Errorf("#%d: unexpected response is received %+v", i, nr)
+		case <-time.After(2 * time.Second):
+		}
+		wStream.CloseSend()
```
is there a close which can close both side?

gyuho replied:
@xiang90  I don't see `Close` method  in `ServerStream`

google.golang.org/grpc/stream.go

```go
// ServerStream defines the interface a server stream has to satisfy.
type ServerStream interface {
	// SendHeader sends the header metadata. It should not be called
	// after SendProto. It fails if called multiple times or if
	// called after SendProto.
	SendHeader(metadata.MD) error
	// SetTrailer sets the trailer metadata which will be sent with the
	// RPC status.
	SetTrailer(metadata.MD)
	Stream
}
```

xiang90 replied:
ok. fine.

gyuho commented:
@xiang90 Feedback addressed, and add two more tests. PTAL. Thanks!

xiang90 commented on integration/cluster_test.go:
```diff
@@ -56,7 +56,7 @@ var (
 	// integration test uses well-known ports to listen for each running member,
 	// which ensures restarted member could listen on specific port again.
-	nextListenPort int64 = 20000
+	nextListenPort int64 = 30000
```
why do we need to change this?

gyuho replied:
Oh thought I revoked it. Will revert it.

xiang90 commented:
LGTM. Thanks for taking a great effort to make this clean.

gyuho commented:
@xiang90 Thanks for feedback!
Will merge after CI passes.

I will create an issue to keep track of other TODO items.
